src/storage/sql_table.cpp

```diff
@@ -1,34 +1,324 @@
 #include "storage/sql_table.h"
 #include <set>
+#include <utility>
 #include <vector>
+
 #include "common/macros.h"
 
 namespace terrier::storage {
 
-std::vector<col_id_t> SqlTable::ColIdsForOids(const std::vector<catalog::col_oid_t> &col_oids) const {
+SqlTable::SqlTable(BlockStore *const store, const catalog::Schema &schema, const catalog::table_oid_t oid)
+    : block_store_(store), oid_(oid) {
+  UpdateSchema(schema);
+}
+
+SqlTable::~SqlTable() {
+  while (tables_.CBegin() != tables_.CEnd()) {
+    auto pair = *(tables_.CBegin());
+    delete (pair.second.data_table);  // Delete the data_table object on the heap
+    tables_.UnsafeErase(pair.first);
+  }
+}
+
+void SqlTable::UpdateSchema(const catalog::Schema &schema) {
+  STORAGE_LOG_DEBUG("Update schema version: {}", uint32_t(schema.GetVersion()));
+  TERRIER_ASSERT(tables_.Find(schema.GetVersion()) == tables_.End(), "schema versions for an SQL table must be unique");
```

Comment:
How do you handle concurrent updates? Also, how do you enforce that layout versions are incremented sequentially? If there are gaps in the schema versions, it may cause your iterator to stop prematurely.

Reply:
Concurrent updates are prevented because the updating must get a write-lock on the table's schema version in the catalog and therefore any subsequent writers will abort because of a write-write conflict.  This approach also guarantees sequential versions because an update will increment their visible version.  Aborted updates won't be visible and pending versions will cause a write-write conflict.